search/query/util_knn.go

```diff
@@ -0,0 +1,121 @@
+//  Copyright (c) 2023 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build vectors
+// +build vectors
+
+package query
+
+import (
+	"context"
+
+	"github.com/blevesearch/bleve/v2/mapping"
+	"github.com/blevesearch/bleve/v2/search"
+	"github.com/blevesearch/bleve/v2/search/searcher"
+	index "github.com/blevesearch/bleve_index_api"
+)
+
+func (q *ConjunctionQuery) Searcher(ctx context.Context, i index.IndexReader, m mapping.IndexMapping, options search.SearcherOptions) (search.Searcher, error) {
+	ss := make([]search.Searcher, 0, len(q.Conjuncts))
+	matchAllSearcherCount := 0
+	knnSearcherCount := 0
+	for _, conjunct := range q.Conjuncts {
+		sr, err := conjunct.Searcher(ctx, i, m, options)
+		if err != nil {
+			for _, searcher := range ss {
+				if searcher != nil {
+					_ = searcher.Close()
+				}
+			}
+			return nil, err
+		}
+		if _, ok := sr.(*searcher.MatchNoneSearcher); ok && q.queryStringMode {
+			// in query string mode, skip match none
+			continue
+		}
+		if _, ok := sr.(*searcher.MatchAllSearcher); ok {
+			matchAllSearcherCount += 1
+		}
+		if _, ok := sr.(*searcher.KNNSearcher); ok {
+			knnSearcherCount += 1
+		}
+		ss = append(ss, sr)
+	}
+
+	if len(ss) < 1 {
+		return searcher.NewMatchNoneSearcher(i)
+	}
+	nctx := context.WithValue(ctx, search.IncludeScoreBreakdownKey, q.retrieveScoreBreakdown)
+
+	// Here, check if it's effectively a single KNN searcher
+	// If so, return new KNN Searcher
+	if knnSearcherCount == 1 && matchAllSearcherCount == 1 {
```

Comment:
Oh this is not complete. What if there're other searchers alongside 1 match_all and 1 knn. This also needs to be applicable ONLY when the operator is `and` here and for disjunction an `or` with match_none.

Reply:
@abhinavdangeti re the second point:
A conjunction query will be formed only when the KNNOperator is "and". Hence, i don't see why there needs to be an explicit check for the operator.
https://github.com/blevesearch/bleve/blob/unstable/search_knn.go#L182-L185